scanpy/get/_aggregated.py

```diff
@@ -0,0 +1,486 @@
+
+    Returns
+    -------
+    pd.DataFrame: Grouped-by Dataframe.
+    """
+    df = df.copy()
+    if key_set is not None:
+        df = df[df[key].isin(key_set)]
+    if pd.api.types.is_categorical_dtype(df[key]):
+        df[key] = df[key].cat.remove_unused_categories()
+    return df.groupby(key).first()[_superset_columns(df, key)]
+
+
+@singledispatch
+def aggregated(
+    adata: AnnData,
+    by: str,
+    func: Union[AggType, Iterable[AggType]],
+    *,
+    dim: Literal['obs', 'var'] = 'obs',
+    weight_key: Optional[str] = None,
+    key_set: Optional[Iterable[str]] = None,
+    dof: int = 1,
+    layer: Optional[str] = None,
+    obsm: Optional[str] = None,
+    varm: Optional[str] = None,
+) -> AnnData:
+    """\
+    Aggregate data based on one of the columns of one of the axes (`obs` or `var`).
+    If none of `layer`, `obsm`, or `varm` are passed in, `X` will be used for aggregation data.
+    If `func` only has length 1 or is just an `AggType`, then aggregation data is written to `X`.
+    Otherwise, it is written to `layers` or `xxxm` as appropriate for the dimensions of the aggregation data.
+
+    Params
+    ------
+    adata
+        :class:`~anndata.AnnData` to be aggregated.
+    by
+        Key of the column to be grouped-by.
+    func
+        How to aggregate.
+    dim
+        Axis on which to find group by column.
+    weight_key
+        Key of the `dim` containing weights for a weighted sum aggregation.
+    key_set
+        Subset of dim on which to filter.
+    dof
+        Degrees of freedom for variance. Defaults to 1.
+    layer
+        If not None, key for aggregation data.
+    obsm
+        If not None, key for aggregation data.
+    varm
+        If not None, key for aggregation data.
+
+    Returns
+    -------
+    Aggregated :class:`~anndata.AnnData`.
+    """
+    data = adata.X
+    write_to_xxxm = None
+    if sum([varm is None, obsm is None, layer is None]) > 1:
+        raise TypeError("Please only provide one (or none) of varm, obsm, or layer")
+    if not varm is None:
+        data = adata.varm[varm]
+        write_to_xxxm = True  # the data will have to be transposed so this is accurate
+    elif not obsm is None:
+        data = adata.obsm[obsm]
+        write_to_xxxm = True
+    elif not layer is None:
+        data = adata.layers[layer]
+        if dim == 'var':
+            data = data.T
+    elif dim == 'var':
+        # i.e., all of `varm`, `obsm`, `layers` are None so we use `X` which must be transposed
+        data = data.T
+    return aggregated(
+        data,
+        groupby_df=getattr(adata, dim),
+        dim=dim,
+        by=by,
+        write_to_xxxm=write_to_xxxm,
+        no_groupby_df=getattr(adata, 'var' if dim == 'obs' else 'obs'),
+        weight_key=weight_key,
+        key_set=key_set,
+        func=func,
+        dof=dof,
+    )
+
+
+@aggregated.register(np.ndarray)
+@aggregated.register(spmatrix)
+def aggregated_from_array(
```

Comment:
As said:

> Is singledispatch the right tool here? Seems like you have a public function operating on AnnData and a private function operating on arrays/matrices. Maybe it should just be that.

I.e. `aggregated` can be a regular function, calling the helper function `aggregated_from_array` with no singledispatch going on.

They have different signatures and only one of them is supposed to be called by users, right?

Reply:
OK, pinging him!